File: addons/math/math.js
console.log('\n- math.js -\n');


var js = {
  pow: function (n, p){
    var r = n,
        p = p - 1,
        i;

    if(arguments.length < 2)
      throw new Error('Wrong number of arguments');

    for(i=0; i < p; i++)
      r*=n;
    
    return r;
  }
};


var test = require('./../../node_modules/bindings')('math');


test.cout();

console.log(test.getString());


console.log(test.add(5, 4));

var bench_pow = 250000000;

console.time('c++');
console.log(test.pow(0.9999999999, bench_pow));
console.timeEnd('c++');

console.time('js');
console.log(js.pow(0.9999999999, bench_pow));
console.timeEnd('js');

console.log('\n- math.js -');

File: addons/math/math.cc
#include <node.h>
#include <v8.h>
#include <iostream>

using namespace v8;
using namespace std;

void Cout(const v8::FunctionCallbackInfo<Value>& args) {
  Isolate* isolate = Isolate::GetCurrent();
  HandleScope scope(isolate);

  cout << "c++ cout\n";
}


void GetString(const v8::FunctionCallbackInfo<Value>& args){
  Isolate* isolate = Isolate::GetCurrent();
  HandleScope scope(isolate);

  args.GetReturnValue().Set(String::NewFromUtf8(isolate, "String!"));
}

void Add(const v8::FunctionCallbackInfo<Value>& args){
  Isolate* isolate = Isolate::GetCurrent();
  HandleScope scope(isolate);

  if(args.Length() < 2){
    isolate->ThrowException(Exception::TypeError(
        String::NewFromUtf8(isolate, "Wrong number of arguments")));
    return;
  }

  double sum = args[0]->NumberValue() + args[1]->NumberValue();
  Local<Number> js_num = Number::New(isolate, sum);

  args.GetReturnValue().Set(js_num);
}

void Pow(const v8::FunctionCallbackInfo<Value>& args){
  Isolate* isolate = Isolate::GetCurrent();
  HandleScope scope(isolate);

  if(args.Length() < 2){
    isolate->ThrowException(Exception::TypeError(
        String::NewFromUtf8(isolate, "Wrong number of arguments")));
    return;
  }

  double n = args[0]->NumberValue();
  double p = args[1]->NumberValue() - 1;
  double r = n;
  
  for(double i = 0; i < p; i++){
    r*=n;
  }

  Local<Number> js_num = Number::New(isolate, r);

  args.GetReturnValue().Set(js_num);
}

void Init(Handle<Object> exports) {
  Isolate* isolate = Isolate::GetCurrent();

  exports->Set(String::NewFromUtf8(isolate, "cout"),
      FunctionTemplate::New(isolate, Cout)->GetFunction());

  exports->Set(String::NewFromUtf8(isolate, "getString"),
      FunctionTemplate::New(isolate, GetString)->GetFunction());

  exports->Set(String::NewFromUtf8(isolate, "add"),
      FunctionTemplate::New(isolate, Add)->GetFunction());

  exports->Set(String::NewFromUtf8(isolate, "pow"),
      FunctionTemplate::New(isolate, Pow)->GetFunction());
}

NODE_MODULE(test, Init)
